presto-main/src/main/java/com/facebook/presto/operator/GroupIdOperator.java

```diff
@@ -0,0 +1,215 @@
+                // first set all grouping columns to true
+                for (int groupingColumn : allGroupingColumns) {
+                    groupingSetNullChannels[i].set(groupingColumn, true);
+                }
+                // then set all the columns in this grouping set to false
+                for (int nonNullGroupingColumn : groupingSetChannels.get(i)) {
+                    groupingSetNullChannels[i].set(nonNullGroupingColumn, false);
+                }
+            }
+
+            return new GroupIdOperator(operatorContext, outputTypes, groupingSetNullChannels);
+        }
+
+        @Override
+        public void close()
+        {
+            closed = true;
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            return new GroupIdOperatorFactory(operatorId, planNodeId, inputTypes, groupingSetChannels);
+        }
+    }
+
+    private final OperatorContext operatorContext;
+    private final List<Type> types;
+    private final BitSet[] groupingSetNullChannels;
+    private final PageBuilder pageBuilder;
+
+    private Page currentPage = null;
+    private int currentPosition = 0;
+    private boolean finishing;
+
+    public GroupIdOperator(
+            OperatorContext operatorContext,
+            List<Type> inputTypes,
+            BitSet[] groupingSetNullChannels)
+    {
+        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+        this.types = requireNonNull(inputTypes, "inputTypes is null");
+        this.pageBuilder = new PageBuilder(types);
+        this.groupingSetNullChannels =  requireNonNull(groupingSetNullChannels, "groupingSetNullChannels is null");
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public List<Type> getTypes()
+    {
+        return types;
+    }
+
+    @Override
+    public void finish()
+    {
+        finishing = true;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return finishing && currentPage == null;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finishing && currentPage == null;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        checkState(!finishing, "Operator is already finishing");
+        checkState(currentPage == null, "currentPage must be null to add a new page");
+
+        currentPage = requireNonNull(page, "page is null");
+        currentPosition = 0;
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        if (currentPage == null && pageBuilder.isEmpty()) {
+            return null;
+        }
+
+        if (currentPage != null) {
+            process();
+        }
+
+        Page output = pageBuilder.build();
+        pageBuilder.reset();
+        return output;
+    }
+
+    private void process()
+    {
+        while (currentPosition < currentPage.getPositionCount() && !pageBuilder.isFull()) {
```

Comment:
this could be more efficient by just processing each groupingSet separately and appending the groupId to the page and nulling out the relevant columns.